File: md2nga.py
# Translate a markdown file to nga.

import argparse
import json
import os
import re

def pass_code(func):
    def wrapper(*args, **kwargs):
        global state
        if state["code"] != "off":
            return args[0]
        res = func(*args, **kwargs)
        return res
    return wrapper

@pass_code
def trans_bidel(line):
    line = re.sub(r'\*\*\*(?P<content>.+?)\*\*\*',
                  '[b][i]\g<content>[/i][/b]',
                  line)
    line = re.sub(r'\*\*(?P<content>.+?)\*\*',
                  '[b]\g<content>[/b]',
                  line)
    line = re.sub(r'\*(?P<content>.+?)\*',
                  '[i]\g<content>[/i]',
                  line)
    line = re.sub(r'~~(?P<content>.+?)~~',
                  '[del]\g<content>[/del]',
                  line)
    return line

@pass_code
def trans_head(line):
    global state
    r = re.match(r'^(?P<level>#+) (?P<head>.*)', line)
    if not r: return line
    if len(r.group("level")) <= state["config"]["head"]:
        line = "[h]{}[/h]".format(r.group("head"))
    elif len(r.group("level")) == state["config"]["head"] + 1:
        line = "[b]{}[/b][h][/h]".format(r.group("head"))
    else:
        line = "[b]{}[/b]".format(r.group("head"))
    return line + "\n"

@pass_code
def trans_image(line):
    return re.sub(r'!\[(?P<content>.*?)\]\((?P<url>[-a-zA-Z0-9@:%_\+.~#?&//=]{2,256}\.[a-z]{2,4}\b(\/[-a-zA-Z0-9@:%_\+.~#?&//=]*)?)\)',
                  '[img]\g<url>[/img]',
                  line)

def trans_url(line):
    return re.sub(r'\[(?P<content>.*?)\]\((?P<url>[-a-zA-Z0-9@:%_\+.~#?&//=]{2,256}\.[a-z]{2,4}\b(\/[-a-zA-Z0-9@:%_\+.~#?&//=]*)?)\)',
                  '[url=\g<url>]\g<content>[/url]',
                  line)

@pass_code
def handle_eof():
    global state
    new_line = ""
    while state["level"]["list"][-1] > 0:
        new_line += " " * (state["level"]["list"][-1] - 2)
        new_line += '[/list]\n'
        state["level"]["list"] = state["level"]["list"][:-1]
    return new_line

@pass_code
def trans_list(line):
    global state
    if re.match(r'(?P<level>(\s+)?)\d+\.', line):
        line = re.sub(r'\d+\.', '-', line, 1)
    p = r'^(?P<level>(\s+)?)[-+*]'
    r = re.match(p, line)
    if not line.strip(): return line
    if not r:
        if state["level"]["list"][-1] == 0: return line
        new_line = handle_eof()
        return new_line + line
    list_level = len(r.group("level")) + 2  # add 2 cuz 0 for no list
    new_line = " " * state["level"]["list"][-1] # start offset
    if list_level > state["level"]["list"][-1]: # new list start
        state["level"]["list"].append(list_level)
        new_line += "[list]" + re.sub(p, '[*]', line, 1)
    elif list_level == state["level"]["list"][-1]:  # list same level
        new_line += re.sub(p, '[*]', line, 1)
    else:   # list back level
        while state["level"]["list"][-1] > list_level:
            new_line = "{}[/list]\n".format(" " * (state["level"]["list"][-1] - 2)) # end offset
            state["level"]["list"] = state["level"]["list"][:-1]
        new_line += (" " * state["level"]["list"][-1]) + re.sub(p, '[*]', line, 1)
    return new_line

def trans_code(line):
    global state
    p = r'\s*?```(?P<language>(\w*)?)'
    r = re.match(p, line)
    if r:
        if state["code"] == "on":
            line = re.sub(r'```((\w*)?)', "[/code]\n", line)
            state["code"] = "to_turn_off"
        else:
            line = re.sub(r'```(?P<language>(\w*)?)', "[code=\g<language>]", line)
            state["code"] = "on"
    return line


def parse_translations(lines):
    trans = []
    for line in lines:
        line = trans_code(line)
        line = trans_bidel(line)
        line = trans_head(line)
        line = trans_image(line)
        line = trans_url(line)
        line = trans_list(line)
        trans.append(line)
        if state["code"] == "to_turn_off":
            state["code"] = "off"
    trans.append(handle_eof())
    return trans

def main(args):
    filename = args.input
    if not os.path.exists(filename):
        raise FileNotFoundError(f'Could not find file "{filename}"')

    with open(filename, 'r', encoding='utf-8') as fp:
        lines = fp.readlines()

    trans = parse_translations(lines)

    if args.output:
        with open(args.output, 'w', encoding='utf-8') as fp:
            fp.write(''.join(trans))
    else:
        for line in trans:
            print(line)

if __name__ == "__main__":
    state = {
        "config":{
            "head": 1,
        },
        "level": {
            "list": [0],
        },
        "code": "off"
    }
    example_usage = ""

    parser = argparse.ArgumentParser(
        description="Translate a markdown file to nga",
        epilog=example_usage,
        formatter_class=argparse.RawDescriptionHelpFormatter)

    parser.add_argument('-i', '--input', help="The markdown file name, e.g. README.md")
    parser.add_argument('-o', '--output', help="If passed, output the nga code to a file, e.g. README.nga")
    parser.add_argument('-hl', '--headlevel', type=int, help="If passed, overwrite the default head level, e.g. 2")

    args = parser.parse_args()

    if not args.input:
        raise parser.error('Must pass an input markdown file.')
    if args.headlevel:
        state["config"]["head"] = args.headlevel

    main(args)

File: docs/js/md2nga.js
"use strict";
let state = {};

function trans_bidel(line){
    line = line.replace(/\*\*\*(.+?)\*\*\*/g, "[b][i]$1[/i][/b]");
    line = line.replace(/\*\*(.+?)\*\*/g, "[b]$1[/b]");
    line = line.replace(/\*(.+?)\*/g, "[i]$1[/i]");
    line = line.replace(/~~(.+?)~~/g, "[del]$1[/del]");
    return line;
}

function trans_head(line){
    let p = /^(#+) (.*)/;
    let r = line.match(p);
    if(!r) return line;
    if (r[1].length <= state["config"]["head"])
        line = "[h]" + r[2] + "[/h]";
    else if (r[1].length == state["config"]["head"] + 1)
        line = "[b]" + r[2] + "[/b][h][/h]";
    else
        line = "[b]" + r[2] + "[/b]";
    return line;
}

function trans_image(line){
    return line.replace(/!\[(?:.*?)\]\(([-a-zA-Z0-9@:%_\+.~#?&//=]{2,256}\.[a-z]{2,4}\b(\/[-a-zA-Z0-9@:%_\+.~#?&//=]*)?)\)/g, '[img]$1[/img]');
}

function trans_url(line){
    return line.replace(/\[(.*?)\]\(([-a-zA-Z0-9@:%_\+.~#?&//=]{2,256}\.[a-z]{2,4}\b(\/[-a-zA-Z0-9@:%_\+.~#?&//=]*)?)\)/g, '[url=$2]$1[/url]');
}

function trans_code(line){
    let p = /\s*?```((\w*)?)/;
    let r = line.match(p);
    if (r){
        if(state["code"] == "on"){
            line = line.replace(/```((\w*)?)/, "[/code]");
            state["code"] = "to_turn_off";
        }else{
            line = line.replace(/```((\w*)?)/, "[code=$1]");
            state["code"] = "on";
        }
    }
    return line;
}

function handle_eof(){
    let new_line = ""
    while(state["level"]["list"][state["level"]["list"].length-1] > 0){
        state["level"]["list"].pop();
        new_line += " ".repeat(Math.max(state["level"]["list"][state["level"]["list"].length-1], 0));
        new_line += '[/list]\n';
    }
    return new_line
}

function trans_list(line){
    let p = /^((\s+)?)\d+\./;
    if (p.test(line))
        line = line.replace(/\d+\./, '-');
    p = /^((\s+)?)[-+*]/;    // $1: level
    let r = line.match(p);
    if (!line.trim()) return line;
    if (!r){
        if(state["level"]["list"][state["level"]["list"].length-1] == 0) return line;
        return handle_eof() + line;
    }
    let list_level = r[1].length + 6;  // add 2 cuz 0 for no list
    let len = state["level"]["list"].length;
    let new_line = " ".repeat(state["level"]["list"][len-1]) // start offset
    if (list_level > state["level"]["list"][len-1]){ // new list start
        state["level"]["list"].push(list_level);
        new_line += "[list]" + line.replace(p, '[*]');
    }else if (list_level == state["level"]["list"][-1]){ //list same level
        new_line += line.replace(p, '[*]');
    }else{   // list back level
        new_line = ""
        while (state["level"]["list"][state["level"]["list"].length-1] > list_level){
            state["level"]["list"].pop();
            new_line += " ".repeat(Math.max(state["level"]["list"][state["level"]["list"].length-1], 0)) + "[/list]\n"; // end offset
        }
        new_line += (" ".repeat(state["level"]["list"][state["level"]["list"].length-1])) + line.replace(p, '[*]');
    }
    return new_line
}

function trans_table(lines) {
    let result = "";
    const origin = lines;
    const length = origin.length;
    for (let i = 0; i < length; i++) {
        if (i + 1 < length &&
            origin[i].startsWith("|") && origin[i].endsWith("|") &&
            origin[i + 1].startsWith("|") && origin[i + 1].endsWith("|") && !/[^\s\|\-\:]/.test(origin[i + 1]) &&
            origin[i + 1].match(/(?<!\\)\|/g).length === origin[i].match(/(?<!\\)\|/g).length) {
            let tablecode = origin[i] + "\n" + origin[i + 1];
            i++;
            for (let j = i + 1; j < length; j++) {
                if (origin[j].startsWith("|") && origin[j].endsWith("|")) {
                    tablecode += "\n" + origin[j];
                    i = j;
                } else {
                    break;
                }
            }
            result += "\n" + trans_table_gen(tablecode);
        } else {
            result += "\n" + origin[i];
        }
    }
    return result.trim().split("\n");
}

function trans_table_gen(tablecode) {
    let trh,
        trd = [];
    const align = [];
    const tablecodelines = tablecode.split(/\n/);
    const delimiterrow = tablecodelines.splice(1, 1);
    delimiterrow[0].substring(1, delimiterrow[0].length - 1).split(/(?<!\\)\|/).map((str) => str.trim()).forEach((delimiter) => {
        let alignside = null;
        if (delimiter.startsWith(":")) {
            alignside = "left";
        }
        if (delimiter.endsWith(":")) {
            if (alignside === "left") {
                alignside = "center";
            } else {
                alignside = "right";
            }
        }
        align.push(alignside);
    });
    tablecodelines.forEach((code, index) => {
        const split = code.substring(1, code.length - 1).split(/(?<!\\)\|/).map((str) => str.trim());
        split.splice(align.length);
        while (split.length < align.length) {
            split.push("");
        }
        if (index === 0) {
            trh = split.map((s, i) => {
                let base = s;
                if (align[i] !== null) {
                    base = `[align=${align[i]}]${s}[/align]`;
                }
                return `[td][b]${base}[/b][/td]`;
            }).join("");
        } else {
            trd.push(split.map((s, i) => {
                let base = s;
                if (align[i] !== null) {
                    base = `[align=${align[i]}]${s}[/align]`;
                }
                return `[td]${base}[/td]`;
            }).join(""));
        }
    });
    return `[table]\n[tr]${trh}[/tr]\n[tr]${trd.join("[/tr]\n[tr]")}[/tr]\n[/table]`;
}
function md2nga(text) {
    let lines = text.split("\n");
    let table_lines = trans_table(lines);
    let trans = Array();
    state = {
        "config":{
            "head": 1,
        },
        "level": {
            "list": [0],
        },
        "code": "off"
    };
    for(let idx in table_lines){
        let line = table_lines[idx];
        line = trans_code(line)
        if (state["code"] == "off"){
            line = trans_bidel(line);
            line = trans_head(line);
            line = trans_image(line);
            line = trans_url(line);
            line = trans_list(line)
        }
        trans.push(line)
        if(state["code"] == "to_turn_off"){
            state["code"] = "off";
        }
    }
    trans.push(handle_eof());
    return trans.join("\n");
}
